examples/ex3/main.cpp

```diff
@@ -0,0 +1,280 @@
+protected:
+  void observableFunction(const double * in, double * out){
+
+    _ffnn->setInput(1, &in[0]);
+    _ffnn->FFPropagate();
+
+    double nnout = _ffnn->getOutput(1);
+    double fx = _ftarget->f(in[0]);
+
+    for(int i=0; i<_nobs; ++i){
+        out[i] = 2.*(nnout - fx)*_ffnn->getVariationalFirstDerivative(1, i);
+      }
+  }
+};
+
+//
+class FitNN1D: public NoisyFunctionWithGradient{
+private:
+  FeedForwardNeuralNetwork * _ffnn;
+  Function1D * _ftarget;
+  MCI * _mcif, * _mcig;
+  long _nmc;
+  double _x0, _step0;
+
+public:
+  FitNN1D(FeedForwardNeuralNetwork * ffnn, Function1D * ftarget, const long &nmc, double * irange): NoisyFunctionWithGradient(ffnn->getNBeta()){
+    _ffnn = ffnn;
+    _ftarget = ftarget;
+    _nmc = nmc;
+
+    _x0 = 0.5*(irange[0] + irange[1]);
+    _step0 = 0.1*(irange[1] - irange[0]);
+
+    _mcif = new MCI(1);
+    _mcig = new MCI(1);
+
+    MCIObservableFunctionInterface * fobs = new NNFitDistance1D(_ffnn, _ftarget);
+    MCIObservableFunctionInterface * gobs = new NNFitGradient1D(_ffnn, _ftarget);
+
+    _mcif->addObservable(fobs);
+    _mcig->addObservable(gobs);
+
+    double ** intrange = new double*[1];
+    intrange[0] = irange;
+    _mcif->setIRange(intrange);
+    _mcig->setIRange(intrange);
+
+    double targetacceptrate = 0.5;
+    _mcif->setTargetAcceptanceRate(&targetacceptrate);
+    _mcig->setTargetAcceptanceRate(&targetacceptrate);
+
+  }
+
+  void f(const double * in, double &f, double &df){
+    int i;
+
+    //set new NN betas
+    for (i=0; i<_ffnn->getNBeta(); ++i){
+      _ffnn->setBeta(i, in[i]);
+    }
+
+    // initial walker position and step size
+    _mcif->setX(&_x0);
+    _mcif->setMRT2Step(&_step0);
+
+    // integrate
+    double * average = new double[_mcif->getNObsDim()];
+    double * error = new double[_mcif->getNObsDim()];
+    _mcif->integrate(_nmc, average, error);
+
+    // write out results
+    f = average[0];
+    df = error[0];
+  }
+
+  void grad(const double * in, double * g, double * dg){
+    int i, nbeta = _ffnn->getNBeta();
+
+    //set new NN betas
+    for (i=0; i<nbeta; ++i){
+      _ffnn->setBeta(i, in[i]);
+    }
+
+    // initial walker position and step size
+    _mcig->setX(&_x0);
+    _mcig->setMRT2Step(&_step0);
+
+    // integrate
+    double * average = new double[_mcig->getNObsDim()];
+    double * error = new double[_mcig->getNObsDim()];
+    _mcig->integrate(_nmc, average, error);
+
+    // write out results
+    for(i=0; i<nbeta; ++i){
+      g[i] = average[i];
+      dg[i] = error[i];
+    }
+  }
+
+};
+
+
+int main() {
+  using namespace std;
+
+  int nl, nh1, nh2;
+  long nmc = 40000l;
+  //double epsx = 0.01;
+  double * irange = new double[2];
+  irange[0] = -10.;
+  irange[1] = 10.;
+
+  cout << "Let's start by creating a Feed Forward Artificial Neural Netowrk (FFANN)" << endl;
+  cout << "========================================================================" << endl;
+  cin.ignore();
```

Comment:
the first times I thought it was a good idea to make an interactive example, asking the user to input the starting parameters, but later on I realised how annoying it can be if you need to run the code multiple times... I strongly recommend not doing that, but rather comment the code and simply output the results.

Reply:
but how am i going to change the parameters? input file? edit the code?
I think asking for this input is adequate and fits the other examples